gtsam/hybrid/GaussianMixture.cpp

```diff
@@ -18,29 +18,78 @@
 
 #include <gtsam/hybrid/GaussianMixture.h>
 #include <gtsam/inference/Conditional.h>
+#include <gtsam/base/utilities.h>
+#include <regex>
+#include <numeric>
 
 namespace gtsam {
 
 // Instantiate base class
 template class Conditional<DCGaussianMixtureFactor, GaussianMixture>;
 
-void GaussianMixture::print(const std::string& s,
-                            const KeyFormatter& keyFormatter) const {
-  BaseFactor::printKeys(s, keyFormatter);
-
-  auto valueFormatter =
-      [](const GaussianFactor::shared_ptr& factor) -> std::string {
-    if (auto conditional =
-            boost::dynamic_pointer_cast<GaussianConditional>(factor))
-      return (boost::format(
-                  "Gaussian conditional on %d frontals given %d parents") %
-              conditional->nrFrontals() % conditional->nrParents())
-          .str();
-    else
-      return "";
+void GaussianMixture::print(const std::string &s,
+                            const KeyFormatter &keyFormatter) const {
+  std::cout << (s.empty() ? "" : s + " ");
+  std::cout << "[";
+
+
+  for (Key key : frontals()) std::cout << keyFormatter(key) << " ";
+  std::cout << "| ";
+  for (Key key : parents()) std::cout << keyFormatter(key) << " ";
+
+  std::cout << "]";
+  std::cout << "{\n";
+
+  auto valueFormatter = [](const GaussianFactor::shared_ptr &v) {
+    auto indenter = [](const GaussianFactor::shared_ptr &p) {
+      RedirectCout rd;
+      p->print();
+      auto contents = rd.str();
+      auto re = std::regex("\n");
+      auto lines =
```

Comment:
Do we need the regex? I don't see why we can't do something much simpler and future-proof.

Reply:
This is very simple and future proof (C++11 -> 20), just a bit slow... But I think this is not a problem for printing, so.

Reply:
But isn't this just reading the print output and then reprinting it with extra indentation? In that case we could simply use the RedirectCout to capture the result of `p->print()` and just return that. That's what I've been doing so far and it works great in terms of readability.

Reply:
Plus it's a lot simpler and less error prone by someone who's not as good at C++ as you :)